tests/logging/test_opentracing.py

```diff
@@ -194,3 +200,80 @@ async def root():
             self._reporter.get_spans(),
             [scopes[1].span, scopes[2].span, scopes[0].span],
         )
+
+    def test_trace_decorator_sync(self) -> None:
+        """
+        Test whether we can use `@trace_with_opname` (`@trace`) and `@tag_args`
+        with sync functions
+        """
+        with LoggingContext("root context"):
+
+            @trace_with_opname("fixture_sync_func", tracer=self._tracer)
+            @tag_args
+            def fixture_sync_func() -> str:
+                return "foo"
+
+            result = fixture_sync_func()
+            self.assertEqual(result, "foo")
+
+        # the span should have been reported
+        self.assertEqual(
+            [span.operation_name for span in self._reporter.get_spans()],
+            ["fixture_sync_func"],
+        )
+
+    def test_trace_decorator_deferred(self) -> None:
+        """
+        Test whether we can use `@trace_with_opname` (`@trace`) and `@tag_args`
+        with functions that return deferreds
+        """
+        reactor = MemoryReactorClock()
+
+        with LoggingContext("root context"):
+
+            @trace_with_opname("fixture_deferred_func", tracer=self._tracer)
+            @tag_args
+            def fixture_deferred_func() -> defer.Deferred[str]:
```

Comment:
Is this the right type?

CI failure ❌ https://github.com/matrix-org/synapse/runs/7697453941?check_suite_focus=true
```
Error: 
Traceback (most recent call last):
  File "/github/workspace/tests/logging/test_opentracing.py", line 236, in test_trace_decorator_deferred
    def fixture_deferred_func() -> defer.Deferred[str]:
builtins.TypeError: 'type' object is not subscriptable

tests.logging.test_opentracing.LogContextScopeManagerTestCase.test_trace_decorator_deferred
```

Reply:
We usually quote `Deferred[T]` in annotations to get around that: `"defer.Deferred[str]"`